### PR TITLE
Enables `use fp32master weights` for ESM2 training

### PR DESCRIPTION
**Pull request by jomitchellnv**
### Description

Enable the option to train using FP32 master/optimizer weights during training. This is useful for FP8  and FP4 low precision training.

#### Usage

You can enable this using `use_fp32_master_weights=True` during training to use FP32 master weights

```python train_fsdp2.py use_fp32_master_weights=True
```

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

#### Triggering Code Rabbit AI Review

To trigger a code review from code rabbit, comment on a pull request with one of these commands:

- @coderabbitai review - Triggers a standard review
- @coderabbitai full review - Triggers a comprehensive review

See https://docs.coderabbit.ai/reference/review-commands for a full list of commands.

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py [hunk 1]
```diff
@@ -96,9 +98,19 @@ def main(args: DictConfig) -> float | None:
...
     transformer_stack = model.esm.encoder.layers if hasattr(model.esm.encoder, "layers") else model.esm.encoder.layer
 
-    for layer in transformer_stack:
-        fully_shard(layer, mesh=device_mesh["dp"])
-    fully_shard(model, mesh=device_mesh["dp"])
+    mp_policy = MixedPrecisionPolicy(
+        param_dtype=torch.bfloat16,  # Cast params to BF16 for forward/backward
+        reduce_dtype=torch.float32,  # Gradient reductions in FP32
+        output_dtype=torch.bfloat16,  # Forward output dtype
+    )
```
Comment (pstjohn) [inline bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py:107]: i think it's safe to always pass along an mp_policy, but parameterize it differently based on the input arguments?

**Review by pstjohn — COMMENTED** [review]
do we need FusedAdam and the various options there to enable fp32 master weights?
can we enable [grad accumulation fusion](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/advanced_optimizations.html#Gradient-accumulation-fusion) in fp32?
maybe it makes sense to try to meet with the TE folks to iron these out?

**Comment by jomitchellnv** [comment]
> do we need FusedAdam and the various options there to enable fp32 master weights? can we enable [grad accumulation fusion](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/advanced_optimizations.html#Gradient-accumulation-fusion) in fp32? maybe it makes sense to try to meet with the TE folks to iron these out?

We don't. FSDP2 gets a copy of the master weights from the model. When you put `params_dtype=<type>` in the model constructor, FSDP will just get the dtype from there. This can be checked using 
```python
print("MIXED PRECISION VERIFICATION (Step X)")
print("="*60)

# 1. Check parameter storage dtype
param = next(model.parameters())
print(f"\n[Parameter Storage]")
print(f"  param.dtype: {param.dtype}")  # Should be FP32

# 2. Check optimizer state dtype
state = optimizer.state.get(param, {})
print(f"\n[Optimizer State]")
print(f"  exp_avg dtype: {state.get('exp_avg', torch.tensor(0)).dtype}")  # Should be FP32
print(f"  exp_avg_sq dtype: {state.get('exp_avg_sq', torch.tensor(0)).dtype}")  # Should be FP32

# 3. Check gradient dtype (computed in BF16)
print(f"\n[Gradients]")
if param.grad is not None:
    print(f"  param.grad.dtype: {param.grad.dtype}")  # Should be BF16

# 4. Check model output dtype (forward pass in BF16)
print(f"\n[Model Output]")
print(f"  outputs.loss.dtype: {outputs.loss.dtype}")  # Should be BF16
if hasattr(outputs, 'logits'):
    print(f"  outputs.logits.dtype: {outputs.logits.dtype}")  # Should be BF16

# Summary
print(f"\n[Summary]")
print(f"  Storage: FP32 = {param.dtype == torch.float32}")
print(f"  Compute: BF16 = {outputs.loss.dtype == torch.bfloat16}")
print(f"  Optimizer: FP32 = {state.get('exp_avg', torch.tensor(0)).dtype == torch.float32}")
print("="*60 + "\n")
```

File: bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py (hunk 1 above)
Comment (jomitchellnv) [reply bionemo-recipes/recipes/esm2_native_te/train_fsdp2.py:107]: Yea it is. I checked here
image
and it looks like we can do None for all of these at default.
